File: rust/tools/pyref/validation.py
"""CLI 和 HTTP API 共用的搜索请求解析与校验。"""

import re
from collections.abc import Mapping

from pyref.models import SEARCH_PROFILES, SearchRequest


_TIME_PATTERN = re.compile(r"^(\d{2}):(\d{2})$")


class RequestValidationError(ValueError):
    def __init__(self, code: str, message: str):
        super().__init__(message)
        self.code = code
        self.message = message


def _value(mapping: Mapping, key: str, default=""):
    value = mapping.get(key, default)
    if isinstance(value, (list, tuple)):
        return value[0] if value else default
    return value


def parse_time(value: str, *, default: int | None = None) -> int:
    value = "" if value is None else str(value).strip()
    if not value:
        if default is None:
            raise RequestValidationError("INVALID_TIME", "时间不能为空")
        return default
    match = _TIME_PATTERN.fullmatch(value)
    if not match:
        raise RequestValidationError("INVALID_TIME", f"无效时间: {value}，应为 HH:MM")
    hour, minute = map(int, match.groups())
    if hour > 23 or minute > 59:
        raise RequestValidationError("INVALID_TIME", f"无效时间: {value}，应为 HH:MM")
    return hour * 60 + minute


def parse_bounded_int(value, name: str, minimum: int, maximum: int, default: int) -> int:
    value = "" if value is None else str(value).strip()
    if not value:
        return default
    try:
        parsed = int(value)
    except (TypeError, ValueError):
        raise RequestValidationError(f"INVALID_{name.upper()}", f"{name} 必须是整数")
    if parsed < minimum or parsed > maximum:
        raise RequestValidationError(
            f"INVALID_{name.upper()}",
            f"{name} 必须在 {minimum}–{maximum} 之间",
        )
    return parsed


def build_search_request(mapping: Mapping) -> SearchRequest:
    """将来自 CLI 或 HTTP 的原始参数转换为 SearchRequest。"""
    from_q = _value(mapping, "from") or _value(mapping, "from_station")
    to_q = _value(mapping, "to") or _value(mapping, "to_station")
    if not from_q or not to_q:
        raise RequestValidationError("MISSING_STATION", "缺少出发站或目的站")

    match_mode = _value(mapping, "match_mode", "fuzzy")
    if match_mode not in ("exact", "fuzzy"):
        raise RequestValidationError("INVALID_MATCH_MODE", "match_mode 必须是 exact 或 fuzzy")

    # 每端独立匹配模式（from_mode/to_mode）：None = 跟随 match_mode
    from_mode = _value(mapping, "from_mode", None) or None
    to_mode = _value(mapping, "to_mode", None) or None
    for name, val in (("from_mode", from_mode), ("to_mode", to_mode)):
        if val is not None and val not in ("exact", "fuzzy"):
            raise RequestValidationError("INVALID_MATCH_MODE", f"{name} 必须是 exact 或 fuzzy")

    profile = _value(mapping, "search_profile", "balanced")
    if profile not in SEARCH_PROFILES:
        raise RequestValidationError("INVALID_SEARCH_PROFILE", "search_profile 必须是 fast/balanced/thorough/complete")

    dep_after = parse_time(_value(mapping, "dep_after"), default=0)
    dep_before = parse_time(_value(mapping, "dep_before"), default=2880)
    arr_after = parse_time(_value(mapping, "arr_after"), default=0)
    arr_before = parse_time(_value(mapping, "arr_before"), default=5760)

    same = parse_bounded_int(
        _value(mapping, "same_transfer"),
        "same_transfer",
        0,
        1440,
        15,
    )
    inter = parse_bounded_int(
        _value(mapping, "inter_transfer"),
        "inter_transfer",
        0,
        1440,
        60,
    )
    max_transfers = parse_bounded_int(
        _value(mapping, "max_transfers"),
        "max_transfers",
        0,
        10,
        3,
    )
    timeout = parse_bounded_int(
        _value(mapping, "timeout"),
        "timeout",
        1,
        600,
        30,
    )

    transfer_city = _value(mapping, "transfer_city") or _value(mapping, "xfer_at") or None

    return SearchRequest(
        from_query=from_q,
        to_query=to_q,
        match_mode=match_mode,
        from_mode=from_mode,
        to_mode=to_mode,
        search_profile=profile,
        earliest_depart=dep_after,
        latest_depart=dep_before,
        earliest_arrive=arr_after,
        latest_arrive=arr_before,
        same_station_transfer_minutes=same,
        interstation_transfer_minutes=inter,
        max_transfers=max_transfers,
        transfer_city_code=transfer_city,
        timeout_seconds=timeout,
    )

File: rust/src/validation.rs
//! 搜索请求参数校验（对齐 Python src/validation.py::build_search_request）。

use crate::models::SearchRequest;
use std::collections::HashMap;

#[derive(Debug, Clone)]
pub struct ValidationError {
    pub code: String,
    pub message: String,
}

/// 解析 HH:MM 为分钟数；空串 → default；格式错误抛 INVALID_TIME（对齐 parse_time）。
pub fn parse_time(value: &str, default: i32) -> Result<i32, ValidationError> {
    let value = value.trim();
    if value.is_empty() {
        return Ok(default);
    }
    let parts: Vec<&str> = value.split(':').collect();
    if parts.len() != 2 {
        return Err(ValidationError {
            code: "INVALID_TIME".into(),
            message: format!("无效时间: {value}，应为 HH:MM"),
        });
    }
    let hour: i32 = match parts[0].parse() {
        Ok(h) => h,
        Err(_) => {
            return Err(ValidationError {
                code: "INVALID_TIME".into(),
                message: format!("无效时间: {value}，应为 HH:MM"),
            });
        }
    };
    let minute: i32 = match parts[1].parse() {
        Ok(m) => m,
        Err(_) => {
            return Err(ValidationError {
                code: "INVALID_TIME".into(),
                message: format!("无效时间: {value}，应为 HH:MM"),
            });
        }
    };
    if hour > 23 || minute > 59 {
        return Err(ValidationError {
            code: "INVALID_TIME".into(),
            message: format!("无效时间: {value}，应为 HH:MM"),
        });
    }
    Ok(hour * 60 + minute)
}

/// 有界整数参数（对齐 parse_bounded_int：空串 → default；非整数/越界抛错）。
pub fn parse_bounded_int(
    value: &str,
    name: &str,
    minimum: i32,
    maximum: i32,
    default: i32,
) -> Result<i32, ValidationError> {
    let value = value.trim();
    if value.is_empty() {
        return Ok(default);
    }
    let parsed: i32 = match value.parse() {
        Ok(v) => v,
        Err(_) => {
            return Err(ValidationError {
                code: format!("INVALID_{}", name.to_uppercase()),
                message: format!("{name} 必须是整数"),
            });
        }
    };
    if parsed < minimum || parsed > maximum {
        return Err(ValidationError {
            code: format!("INVALID_{}", name.to_uppercase()),
            message: format!("{name} 必须在 {minimum}–{maximum} 之间"),
        });
    }
    Ok(parsed)
}

/// 将 HTTP 查询参数转换为 SearchRequest（对齐 build_search_request）。
pub fn build_search_request(params: &HashMap<String, String>) -> Result<SearchRequest, ValidationError> {
    let get = |keys: &[&str]| -> String {
        for k in keys {
            if let Some(v) = params.get(*k) {
                if !v.is_empty() {
                    return v.clone();
                }
            }
        }
        String::new()
    };

    let from_q = get(&["from", "from_station"]);
    let to_q = get(&["to", "to_station"]);
    if from_q.is_empty() || to_q.is_empty() {
        return Err(ValidationError {
            code: "MISSING_STATION".into(),
            message: "缺少出发站或目的站".into(),
        });
    }

    let match_mode = get(&["match_mode"]);
    let match_mode = if match_mode.is_empty() { "fuzzy".to_string() } else { match_mode };
    if match_mode != "exact" && match_mode != "fuzzy" {
        return Err(ValidationError {
            code: "INVALID_MATCH_MODE".into(),
            message: "match_mode 必须是 exact 或 fuzzy".into(),
        });
    }

    let from_mode = get(&["from_mode"]);
    let to_mode = get(&["to_mode"]);
    for (name, val) in [("from_mode", from_mode.as_str()), ("to_mode", to_mode.as_str())] {
        if !val.is_empty() && val != "exact" && val != "fuzzy" {
            return Err(ValidationError {
                code: "INVALID_MATCH_MODE".into(),
                message: format!("{name} 必须是 exact 或 fuzzy"),
            });
        }
    }

    let profile = get(&["search_profile"]);
    let profile = if profile.is_empty() { "balanced".to_string() } else { profile };
    if !["fast", "balanced", "thorough", "complete"].contains(&profile.as_str()) {
        return Err(ValidationError {
            code: "INVALID_SEARCH_PROFILE".into(),
            message: "search_profile 必须是 fast/balanced/thorough/complete".into(),
        });
    }

    let dep_after = parse_time(&get(&["dep_after"]), 0)?;
    let dep_before = parse_time(&get(&["dep_before"]), 2880)?;
    let arr_after = parse_time(&get(&["arr_after"]), 0)?;
    let arr_before = parse_time(&get(&["arr_before"]), 5760)?;
    let same = parse_bounded_int(&get(&["same_transfer"]), "same_transfer", 0, 1440, 15)?;
    let inter = parse_bounded_int(&get(&["inter_transfer"]), "inter_transfer", 0, 1440, 60)?;
    let max_transfers = parse_bounded_int(&get(&["max_transfers"]), "max_transfers", 0, 10, 3)?;
    let timeout = parse_bounded_int(&get(&["timeout"]), "timeout", 1, 600, 30)?;

    let transfer_city = get(&["transfer_city", "xfer_at"]);

    Ok(SearchRequest {
        from_query: from_q,
        to_query: to_q,
        match_mode,
        from_mode: if from_mode.is_empty() { None } else { Some(from_mode) },
        to_mode: if to_mode.is_empty() { None } else { Some(to_mode) },
        search_profile: profile,
        earliest_depart: dep_after,
        latest_depart: dep_before,
        earliest_arrive: arr_after,
        latest_arrive: arr_before,
        same_station_transfer_minutes: same,
        interstation_transfer_minutes: inter,
        max_transfers: max_transfers as usize,
        transfer_city_code: if transfer_city.is_empty() { None } else { Some(transfer_city) },
        timeout_seconds: timeout as u64,
    })
}
